dxr/plugins/descriptor/__init__.py

```diff
@@ -0,0 +1,110 @@
+        return []
+
+
+class FileToIndex(dxr.indexers.FileToIndex):
+    """Do lots of work to yield a description needle."""
+
+    comment_re = re.compile(r'(?:.*?/\*+)(?:\s*\*?\s*)(?P<description>.*?)(?:(?:\*+/.*)|(?:$))', flags=re.M)
+    docstring_re = re.compile(r'(\'\'\'|""")(?:\s*)(?P<description>.*?)(?:(?:(\'\'\'|"""))|(?:$))', flags=re.M)
+    title_re = re.compile(r'<title>([^<]*)</title>')
+
+    def needles(self):
+        if self.contains_text():
+            extension = splitext(self.path)[1]
+            description = None
+            if extension:
+                try:
+                    # Find the describer method, skipping the dot on extension.
+                    describer = getattr(self, 'describe_' + extension[1:])
+                except AttributeError:
+                    # Don't have a descriptor function for this file type, we can try generic later.
+                    pass
+                else:
+                    description = describer()
+            if not description:
+                description = self.generic_describe()
+            if description:
+                yield 'Description', description[:100].strip()
+
+    def describe_html(self):
+        """Return the contents of the <title> tag."""
+
+        match = re.search(self.title_re, self.contents)
+        if match:
+            return match.group(1)
+
+    def describe_py(self):
+        """Return the contents of the first line of the first docstring if
+        there is one in the first 60 lines."""
+
+        sixty_lines = self.contents.splitlines(True)[:60]
+        match = re.search(self.docstring_re, ''.join(sixty_lines))
+        if match:
+            return match.group('description')
+
+    def generic_describe(self):
+        """Look at the first 60 lines for a match for {{self.path|description}}
+        [delimiter] text, and return the first text we find. Unless it's a
+        readme, then return the first line."""
+
+        sixty_lines = self.contents.splitlines(True)[:60]
+        filename = basename(self.path)
+        # Is it a readme? Just return the first non-empty line.
+        if "readme" in filename.lower():
+            for line in sixty_lines:
+                if line:
+                    return line
+        # Not a readme file, try to match the filename: description pattern.
+        root, ext = splitext(filename)
+        delimiters = ':,-'
+        try:
+            description_re = re.compile(r'({}|{}|description)\
+                                         ({})?\s*([{}]\n?)\s*\
+                                         (?P<description>[\w\s-]+)'.format(self.path,
+                                                                           root,
+                                                                           ext,
+                                                                           delimiters),
+                                        re.IGNORECASE)
+            for line in sixty_lines:
```

Comment:
mxr/Local.pm:
"# Read in the beginning of the file into a string. I chose 60 because the
# Berkeley copyright notice is around 40 lines long so we need a bit more
# than this."